src/io/ioexternalbus.cpp

```diff
@@ -28,10 +50,217 @@ QString IOExternalBus::description() const {
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+
+    dprintf("===============>ioRead off=%lx  size=%x  value=%x\n", offset, size, rvalue);
+    return rvalue;
+}
+
+void IOExternalBus::ioWrite(AInt offset, VInt value, unsigned size) {
+    dprintf("===============>ioWrite off=%lx size=%lx  value=%x\n", offset, size, value);
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload[2];
+        payload[0] = htonl(offset);
+        payload[1] = htonl(value);
+        QByteArray dp = QByteArray(static_cast<char*>((void*)payload), 8);
+        if (send_cmd(VB_PWRITE, 8, &dp) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(cmd_header) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        if (cmd_header.msg_type == VB_PWRITE) {
+            dprintf("write ok\n");
+        } else {
+            skt_use = 0;
+            printf("write error !!!!!!!!!!!!!!!!!!!!!!\n");
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+}
+
+void IOExternalBus::connectButtonTriggered() {
+    if (m_ui->connectButton->text().contains("Connect")) {
+        tcpSocket->abort();
+        if (tcpSocket->connectToHost(m_ui->address->text(), m_ui->port->value())) {
+            if (send_cmd(VB_PINFO) < 0) {
+                disconnectOnError();
+                return;
+            }
+            cmd_header_t cmd_header;
+
+            if (recv_cmd(cmd_header) < 0) {
+                disconnectOnError();
+                return;
+            }
+            if (cmd_header.payload_size) {
+                QByteArray buff(cmd_header.payload_size + 1, 0);
+                if (recv_payload(buff, cmd_header.payload_size) < 0) {
+                    disconnectOnError();
+                    return;
+                }
+                dprintf("%s\n", buff.data());
+
+                QJsonParseError error;
+                QJsonDocument desc = QJsonDocument::fromJson(buff.data(), &error);
+                if (error.error == QJsonParseError::NoError) {
+                    const unsigned int addrw = desc.object().value(QString("address width")).toInt();
+                    QJsonObject osymbols = desc.object().value(QString("symbols")).toObject();
+
+                    m_regDescs.clear();
+                    for (QJsonObject::iterator i = osymbols.begin(); i != osymbols.end(); i++) {
+                        m_regDescs.push_back(
+                            RegDesc{i.key(), RegDesc::RW::RW, addrw * 8, static_cast<AInt>(i.value().toInt()), true});
+                    }
+
+                    ByteSize = addrw * osymbols.count();
+
+                    m_ui->connectButton->setText("Disconnect");
+                    m_ui->status->setText("Connected");
```

Comment:
This pattern appears three times in the function. I think it could be good to have a local variable `ConnectionStatus m_status` with a `ConnectionStatus` enum. Then, we'd have a function `updateConnectionStatus()` that `switch(m_status)...` sets the UI components to reflect what is going on.

This would also allow you to remove these `m_ui` calls from here, replace them with a single assignment of the current status, and at the end of this function, a call to `updateConnectionStatus`.

Reply:
Done.